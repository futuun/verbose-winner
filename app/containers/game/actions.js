import types from './actionTypes'

export function letterMissed(key) {
  return {
    type: types.letterMissed,
    key,
  }
}

export function letterExposed(key) {
  return {
    type: types.letterExposed,
    key,
  }
}

function newWord(word) {
  return {
    type: types.newWord,
    word,
  }
}

export function fetchNewWord() {
  return (dispatch) => {
    dispatch({ type: types.wordFetching })
    const uri = 'http://api.wordnik.com:80/v4/words.json/randomWord' +
      '?hasDictionaryDef=true' +
      '&includePartOfSpeech=noun' +
      '&minCorpusCount=0' +
      '&maxCorpusCount=-1' +
      '&minDictionaryCount=1' +
      '&maxDictionaryCount=-1' +
      '&minLength=5' +
      '&maxLength=11' +
      '&api_key=a2a73e7b926c924fad7001ca3111acd55af2ffabf50eb4ae5'

    return fetch(uri)
      .then((res) => {
        if (res.status >= 400) {
          throw new Error(res.status)
        }
        return res.json()
      })
      .then(res => res.word)
      .then(word => dispatch(newWord(word)))
      .catch(() => dispatch({ type: types.wordFetchingError }))
  }
}
