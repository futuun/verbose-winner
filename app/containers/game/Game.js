import React, { Component, PropTypes } from 'react'
import { connect } from 'react-redux'
import styles from './styles.sass'
import * as actionsToBind from './actions'
import Folk from '../../components/folk'
import MissedLetters from '../../components/missedLetters'

export class Game extends Component {
  /**
   * Will dispatch char if pressed key was letter.
   * Character will always be uppercase.
   *
   * @param {Event} e An event
   * @return {string} Character if it was letter or null
   * @memberOf Game
   */
  static getLetterFromKeyEvent(e) {
    const charCode = e.which || e.keyCode

    if (charCode >= 65 && charCode <= 90) {
      return String.fromCharCode(charCode)
    } else if (charCode >= 97 && charCode <= 122) {
      return String.fromCharCode(charCode - 32)
    }

    return null
  }

  constructor(props) {
    super(props)

    this.checkIfCharExist = ::this.checkIfCharExist
    this.handleKeyUpEvent = ::this.handleKeyUpEvent
  }

  componentWillMount() {
    document.addEventListener('keyup', this.handleKeyUpEvent, true)
    this.props.fetchNewWord()
  }

  handleKeyUpEvent(e) {
    if (this.props.data.theEnd) return null

    const letter = Game.getLetterFromKeyEvent(e)
    if (letter) {
      return this.checkIfCharExist(letter)
    }

    return null
  }

  /**
   * Checks if character should be addeded to any array.
   *
   * @param {string} char A test character
   * @returns {function|false} Function if should be added
   * @memberOf Game
   */
  checkIfCharExist(char) {
    const {
      data: {
        missedLetters,
        exposedLetters,
        secretWord,
      },
      letterExposed,
      letterMissed,
    } = this.props

    if (secretWord.includes(char)) {
      if (!exposedLetters.includes(char)) {
        return letterExposed(char)
      }
    } else if (!missedLetters.includes(char)) {
      return letterMissed(char)
    }

    return false
  }

  render() {
    const {
      exposedLetters,
      secretWord,
      missedLetters,
    } = this.props.data

    const zoom = window.innerWidth < 1920
      ? -1920 + window.innerWidth
      : window.innerWidth - 1920

    return (
      <div className={styles.game}>
        <div className={styles.board} style={{ transform: `translate3d(0, 0, ${zoom}px)` }}>
          <div className={styles.missed}>
            <Folk missedLetters={missedLetters} />
            <MissedLetters letters={missedLetters} />
          </div>
          {JSON.stringify(secretWord || '')}<br />
          {JSON.stringify(exposedLetters || '')}<br />
        </div>
      </div>
    )
  }
}

Game.propTypes = {
  letterExposed: PropTypes.func.isRequired,
  letterMissed: PropTypes.func.isRequired,
  fetchNewWord: PropTypes.func.isRequired,
  data: PropTypes.shape({
    missedLetters: PropTypes.arrayOf(PropTypes.string),
    exposedLetters: PropTypes.arrayOf(PropTypes.string),
    secretWord: PropTypes.string,
    theEnd: PropTypes.bool,
  }).isRequired,
}

function select(state) {
  return {
    data: state.game,
  }
}

export default connect(select, actionsToBind)(Game)
