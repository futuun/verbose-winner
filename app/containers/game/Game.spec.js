import React from 'react'
import renderer from 'react-test-renderer'
import TestUtils from 'react-addons-test-utils'
import { Game } from './Game'
import * as actions from './actions'

const initialState = {
  missedLetters: ['A'],
  exposedLetters: ['Y'],
  wordFetching: false,
  secretWord: 'KOPYTKO',
}

test('Initial state snapshot', () => {
  fetch.mockResponse(JSON.stringify({ word: 'kopytko' }))

  const component = renderer.create(
    <Game data={initialState} {...actions} />
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

describe('Game: theEnd set to true (returns null for every input)', () => {
  const view = TestUtils.renderIntoDocument(
    <Game {...actions} data={{ theEnd: true }} />
  )

  it('Key `s`', () => {
    const result = view.handleKeyUpEvent({ key: 's', keyCode: 83, which: 83 })
    expect(result).toBeNull()
  })

  it('Empty call', () => {
    const result = view.handleKeyUpEvent()
    expect(result).toBeNull()
  })
})

describe('Game: handleKeyUpEvent', () => {
  const view = TestUtils.renderIntoDocument(
    <Game
      {...actions}
      letterExposed={() => 'letterExposed'}
      letterMissed={() => 'letterMissed'}
      fetchNewWord={() => 'fetchNewWord'}
      data={initialState}
    />
  )

  it('should pass S to letterMissed', () => {
    const result = view.handleKeyUpEvent({ key: 's', keyCode: 83, which: 83 })
    expect(result).toEqual('letterMissed')
  })

  it('Should do nothing', () => {
    const result = view.handleKeyUpEvent({ key: 'shift', keyCode: 16, which: 16 })
    expect(result).toBeNull()

    const result2 = view.handleKeyUpEvent('y')
    expect(result2).toBeNull()

    const result3 = view.handleKeyUpEvent('a')
    expect(result3).toBeNull()

    const result4 = view.handleKeyUpEvent({ key: 'A', keyCode: 97, which: 97 })
    expect(result4).toEqual(false)
  })

  it('Should add to missed', () => {
    const result = view.handleKeyUpEvent({ key: 'q', keyCode: 81, which: 81 })
    expect(result).toEqual('letterMissed')
  })

  it('Should add to exposed', () => {
    const result = view.handleKeyUpEvent({ key: 'o', keyCode: 79, which: 79 })
    expect(result).toEqual('letterExposed')
  })
})

describe('Game: getLetterFromKeyEvent', () => {
  it('Should return `S`', () => {
    const result = Game.getLetterFromKeyEvent({ key: 's', keyCode: 83, which: 83 })
    expect(result).toEqual('S')
  })

  it('Should return `A`', () => {
    const result = Game.getLetterFromKeyEvent({ key: 'A', keyCode: 97, which: 97 })
    expect(result).toEqual('A')
  })

  it('Should return null', () => {
    const result = Game.getLetterFromKeyEvent({ key: 'Shift', keyCode: 16, which: 16 })
    expect(result).toBeNull()
  })
})

describe('Game: checkIfCharExist', () => {
  const view = TestUtils.renderIntoDocument(
    <Game
      {...actions}
      letterExposed={() => 'letterExposed'}
      letterMissed={() => 'letterMissed'}
      fetchNewWord={() => 'fetchNewWord'}
      data={initialState}
    />
  )

  it('Shouldn\'t add', () => {
    const result = view.checkIfCharExist('A')
    expect(result).toEqual(false)

    const result2 = view.checkIfCharExist('Y')
    expect(result2).toEqual(false)
  })

  it('Should add to missed', () => {
    const result = view.checkIfCharExist('Q')
    expect(result).toEqual('letterMissed')
  })

  it('Should add to exposed', () => {
    const result = view.checkIfCharExist('O')
    expect(result).toEqual('letterExposed')
  })
})
