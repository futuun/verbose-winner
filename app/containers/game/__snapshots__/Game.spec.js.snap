exports[`test Initial state snapshot 1`] = `
<div
  className="game">
  <div
    className="board"
    style={
      Object {
        "transform": "translate3d(0, 0, -896px)",
      }
    }>
    <div
      className="missed">
      <div
        className="folk">
        <div
          className="img bar" />
        <div
          className="img head" />
      </div>
      <div
        className="missedLetters">
        <div
          className="title">
          YOU MISSED:
        </div>
        <div
          className="letters">
          <span>
            A
          </span>
        </div>
      </div>
    </div>
    \"KOPYTKO\"
    <br />
    [\"Y\"]
    <br />
  </div>
</div>
`;
